lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,306 @@
+class SpellCheckerManager
+  checkers: []
+  locales: []
+  localePaths: []
+  useLocales: false
+  localeCheckers: null
+  knownWords: []
+  addKnownWords: false
+  knownWordsChecker: null
+
+  addPluginChecker: (checker) ->
+    console.log "spell-check: addPluginChecker:", checker
+    @addSpellChecker checker
+  addSpellChecker: (checker) ->
+    console.log "spell-check: addSpellChecker:", checker
+    @checkers.push checker
+
+  removeSpellChecker: (spellChecker) ->
+    @checkers = @checkers.filter (plugin) -> plugin isnt spellChecker
+
+  check: (args, text) ->
+    # Make sure our deferred initialization is done.
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
```

Comment:
Now I have opinions about the word API. I banged up a quick [gist](https://gist.github.com/dmoonfire/eb9cc49a2729f78820f2) which has a couple implementations with a low-resolution timing (Date - Date). Mainly, it creates 100k words via `lorem-ipsum` and then puts it through four different combinations:

1. The entire buffer in one shot. (493 ms)
2. Each words, one at a time. (989 ms)
3. With a small `lru-cache` (1535 ms)
5. With an unlimited life cache. (45 ms)

Now, the `lru-cache` is probably implementation-specific, but it gives a rough idea of caching the results of the more common words being checked (once "a" is correct, "a" will always be right).

For the unlimited cache: Using my own novel as an example, I grabbed the first ten chapters which came out to 22k words but only 3.9k distinct words. Storing the last 5k words in memory and then wiping the cache if it exceeded it would allow you to probably easily use a fast cache and take advantage of that 45 ms lookup for the entire buffer. I'd probably say any word over 20 characters is always rechecked and skips the cache to avoid polluting it with obscure words.

The idea of the cache is that most languages have a certain redundancy (articles, prepositions, etc) that fills in most of the language.

In the worst case, checking every word on 100k words is about double what it is to send the entire buffer as a single string.

An advantage of using a word-by-word is that we could eventually have a hook that lets a grammar give us a different rule for tokenization, which would allow us to do PascalCase and camelCase (https://github.com/atom/spell-check/issues/91).

I think with a small cache (5k words, settable via config), it would be advantageous to switch to a word-by-word. It would let us drop the dependency on `multi-integer-range` and simplify the logic for marking up the errors. If you are okay with that, I'll rework the implementation and push it back up. If so, do you want it as a delta against this one or rebase it off the head as a single commit? (I'd squash everything before asking you for the final pull.)